opencog/attention/experimental/ForgettingAgent.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * opencog/attention/ForgettingAgent.cc
+ *
+ * Copyright (C) 2008 by OpenCog Foundation
+ * Written by Joel Pitt <joel@fruitionnz.com>
+ * All Rights Reserved
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <algorithm>
+#include <sstream>
+
+#define DEPRECATED_ATOMSPACE_CALLS
+#include <opencog/atomspace/AtomSpace.h>
+
+#include <opencog/cogserver/server/Agent.h>
+#include <opencog/cogserver/server/CogServer.h>
+#include <opencog/cogserver/server/Factory.h>
+#include <opencog/attention/atom_types.h>
+#include <opencog/util/Config.h>
+#include "ForgettingAgent.h"
+
+using namespace opencog;
+
+ForgettingAgent::ForgettingAgent(CogServer& cs) :
+    Agent(cs)
+{
+    std::string defaultForgetThreshold;
+    std::ostringstream buf;
+
+    // No limit to lti of removed atoms
+    // Convert MAXLTI to a string for storing in the configuration
+    buf << AttentionValue::MAXLTI;
+    defaultForgetThreshold = buf.str();
+    config().set("ECAN_FORGET_THRESHOLD",defaultForgetThreshold);
+
+    forgetThreshold = (AttentionValue::lti_t)
+                      (config().get_int("ECAN_FORGET_THRESHOLD"));
+
+    //Todo: Make configurable
+    maxSize = config().get_int("ECAN_ATOMSPACE_MAXSIZE");
+    accDivSize = config().get_int("ECAN_ATOMSPACE_ACCEPTABLE_SIZE_SPREAD");
+
+    // Provide a logger, but disable it initially
+    log = NULL;
+    setLogger(new opencog::Logger("ForgettingAgent.log", Logger::WARN, true));
+}
+
+ForgettingAgent::~ForgettingAgent()
+{
+    if (log) delete log;
+}
+
+Logger* ForgettingAgent::getLogger()
+{
+    return log;
+}
+
+void ForgettingAgent::setLogger(Logger* _log)
+{
+    if (log) delete log;
+    log = _log;
+}
+
+void ForgettingAgent::run()
+{
+    log->fine("=========== ForgettingAgent::run =======");
+    a = &_cogserver.getAtomSpace();
+    forget();
+}
+
+void ForgettingAgent::forget()
+{
+    HandleSeq atomsVector;
+    std::back_insert_iterator<HandleSeq> output2(atomsVector);
+    int count = 0;
+    int removalAmount;
+    bool recursive;
```

Comment:
The standard C++ convention is to only declare variables when you use them, instead of declaring them at the top of the subroutine.  The need to declare variables at the top of a routine was a compiler limitation dating back to the 1980's, and was fixed in most compilers in the 1990's.  Its not a useful style, and makes it harder to read the code.  (This is also where the underbar convention is useful: so, for example `bool recursive` makes it clear that recursive is a local variable, scoped to the local block, whereas `_recursive` would indicate an object-scope variable.